ASN1 FUZZER

dannyallover opened the pull request.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+                                    const size_t len_pos) {
+  uint8_t len_bytes = GetNumBytes(assigned_len);
+  if (assigned_len > 127) {
+    uint8_t longForm = (1 << 7);
+    longForm += len_bytes;
+    AppendBytes(longForm, len_pos);
```
```suggestion
    uint8_t long_form = 0x80 | len_bytes
    AppendBytes(long_form, len_pos);
```

Or you could even collapse that further.

dannyallover replied:
I explicitly define long_form to make it clear what I'm passing in. If we can assume, that the reader should understand what we're doing here, then I can take out a lot of these middle steps.

sleevi replied:
Yeah. I think this would be sufficient to add comments referencing the portion of the spec that this comes from, and a brief description.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+                                       const size_t len_pos) {
+  size_t assigned_len = 0;
+  if (len.has_length_override()) {
+    sscanf(len.length_override().c_str(), "%zu", &assigned_len);
```
We definitely *don't* want to use sscanf.

I think this might require rethinking how this function works in the broader context: the idea is that you'd directly copy the bytes from `length_override` for where the `length` would normally go.

So you'd just
```suggestion
    encoder_.insert(encoder_.begin() + len_pos, len.length_override().begin(), len.length_override().end());
    return len.length_override().size();
```

dannyallover replied:
I need to convert it to size_t because I need to eventually pass it through EncodeLongForm. If sscanf needs to be avoided, I can do something like:
```
  std::stringsteam len_read(len.length_override());
  len_read >> assigned_len;
```

sleevi replied:
I'm not sure I understand why it needs to go through `EncodeLongForm`? I think the goal of avoiding sscanf is to try to avoid any parsing of the input; it should be something we can just splat into the buffer, like above. I'm probably missing an important detail here, we can chat about it in person :)

dannyallover replied:
Ah that's right, we're just splatting length_override in the buffer -- not trying to properly encode it. Yes, I follow and agree now.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+        len = 0;
+      }
+    }
+  }
```
```suggestion
  size_t len = 0;
  for (const auto &val_ele : val.val_array()) {
    if (val_ele.has_pdu()) {
      len += ParsePDU(val_ele.pdu());
      continue;
    } 
    len = val_ele.val_bits().size();
    encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
                      val_ele.val_bits().end());
}
```

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,45 @@
+}
+
+message ValueElement {
+  oneof types {
+    PDU pdu = 1;
+    bytes val_bits = 2;
+  }
```
I *think* this can collapse into

```suggestion
  optional PDU pdu = 1;
  required bytes val_bits = 2;
```

It's possible for `val_bits` to be present, but for the actual bytes to be `len == 0` (e.g. for a ASN.1 `NULL` type), and this should simplify the logic later.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+  return len;
+}
+
+size_t ASN1ProtoConverter::HighTagForm(const uint8_t cls, const uint8_t enc,
```
```suggestion
uint64_t ASN1ProtoConverter::HighTagForm(const uint8_t cls, const uint8_t enc,
```

We should distinguish when you're using a 64-bit integer explicitly, and when you're representing a size of bytes / offset. This probably will have cascading effects, as I think you've got a few overloaded functions here related.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+  size_t id_parsed = (cls | enc | 0x1F);
+  id_parsed <<= 8;
+  for (uint8_t i = numBytes; i != 0; i--) {
+    id_parsed |= ((0x00 << 7) | ((tag >> (i * 7)) & 0x7F));
```
```suggestion
    id_parsed |= (tag >> (i * 7)) & 0x7F;
```

The 0x00 << 7 isn't needed because of your 0x7F mask

sleevi commented on projects/asn1-proto/asn1_proto_converter.h:
```diff
@@ -0,0 +1,33 @@
+private:
+  std::stringstream der_;
+  std::vector<uint8_t> encoder_;
+  size_t ParsePDU(const PDU &pdu);
+  size_t ParseIdentifier(const Identifier &id);
+  size_t HighTagForm(const uint8_t cl, const uint8_t enc, const uint32_t tag);
+  size_t ParseLength(const Length &len, size_t actual_len, size_t len_pos);
+  size_t LongForm(size_t assigned_len, size_t len_pos);
+  size_t ParseValue(const Value &val);
```
Suggestion: I would suggest naming these "Encode" or "Append" as opposed to "Parse", since that's actually what you're doing.

sleevi commented on projects/asn1-proto/asn1_proto_converter.h:
```diff
@@ -0,0 +1,33 @@
+  std::vector<uint8_t> encoder_;
+  size_t ParsePDU(const PDU &pdu);
+  size_t ParseIdentifier(const Identifier &id);
+  size_t HighTagForm(const uint8_t cl, const uint8_t enc, const uint32_t tag);
```
```suggestion
  size_t EncodeHighTagForm(const uint8_t cl, const uint8_t enc, const uint32_t tag);
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.h:
```diff
@@ -0,0 +1,33 @@
+  size_t ParseIdentifier(const Identifier &id);
+  size_t HighTagForm(const uint8_t cl, const uint8_t enc, const uint32_t tag);
+  size_t ParseLength(const Length &len, size_t actual_len, size_t len_pos);
+  size_t LongForm(size_t assigned_len, size_t len_pos);
```
```suggestion
  size_t EncodeLongForm(size_t assigned_len, size_t len_pos);
```

dannyallover commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,134 @@
+
+namespace asn1_proto {
+
+uint8_t ASN1ProtoConverter::GetNumBytes(size_t num) {
```
Fixed bug here.

dannyallover commented:
PTAL @sleevi :)

dannyallover commented:
Update: I fixed the bugs that I found. Also, I changed Tag to have either a uint32 random tag or a tag that we know exists. While just having an uint32 for the tag will effectively reach the same outcome, this change will allow the fuzzer to obtain coverage much quicker.

sleevi commented:
Have you pushed those changes?

I’m concerned about a specific list of tags, as mentioned yesterday, so I’m curious to see what results.

dannyallover commented:
> Have you pushed those changes?
> 
> I’m concerned about a specific list of tags, as mentioned yesterday, so I’m curious to see what results.

I did (https://github.com/dannyallover/oss-fuzz/tree/dhalawi/asn1/proto_to_der_2/projects/asn1-proto). Hmm, I wonder why they are not reflecting on this PR.

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,79 @@
+  required KnownTag known_tag = 2;
+}
+
+enum KnownTag {
+  bool = 1;
+  int = 2;
+  bit_string = 3;
+  octet_string = 4;
+  null = 5;
+  object_identifier = 6;
+  object_descriptor = 7;
+  instance_of = 8; // or external
+  real = 9;
+  enumerated = 10;
+  embedded_pdv = 11;
+  utf8_string = 12;
+  relative_oid = 13;
+  sequence = 16; // or sequence of
+  set = 17; // or set of
+  numeric_string = 18;
+  printable_string = 19;
+  teletex_string = 20; // or t61 string
+  videotex_string = 21;
+  ia5_string = 22;
+  utc_time = 23;
+  generalized_time = 24;
+  graphic_string = 25;
+  visible_string = 26;
+  general_string = 27;
+  universal_string = 28;
+  character_string = 29;
+  bmp_string = 30;
```
As I mentioned yesterday, I'm not sure we necessarily want this. This is something that can be easily addressed by a seed corpus for any fuzzer, and it's also not something general. That is, because this is only applicable for UNIVERSAL class items, it doesn't really have the same semantic meaning for other classes, so it makes this proto a little more confusing.

@dor1s would be able to advise on this. My understanding is that a seed corpus filled with, say, tags in this range, should be able to provide the same accelerated view.

dannyallover replied:
Hmm, my understanding was that the tag is going to be constant (that is an integer is always going to have tag 2), and then Class (universal, application, etc) is determined by the 7th and 8th high bit? Let me know if I'm off here.

If this can be easily addressed in a seed corpus, then I would be in agreement with you. Let me look into it, as well as get @Dor1s' input

sleevi replied:
@dannyallover Right, but that doesn't make sense with this protobuf.

Put differently, you've overloaded `KnownTag` to mean complete `Identifier`s, which makes it tricky because you put `KnownTag` within `Tag`. An `OCTET STRING` is not just a `Tag.random_tag` == 4, but is in fact a
```
Identifier {
  Class = Universal,
  Encoding = <random>
  Tag = 4
}
```

If that makes any more sense? Like, you could do `KnownIdentifier` to represent this (although then that introduces complexity with `Encoding`, which can vary for these tags), but like, the following doesn't make any sense:
```
Identifier {
  Class = Context-Specific
  Encoding = <random>
  Tag = OCTET STRING
}
```

That's because, in that context (that is, `Class` == `Context-Specific`), the literal value `4` can have different semantic meaning than `OCTET STRING` - it means the literal 4th tag.

But this is where the seed corpus can/will help. The fuzzer will quickly discover the value of low-value tags here and find them, irrespective of a seed corpus, but the seeding of values here will also help for what a "PDU" can be.

Dor1s replied:
Using a seed corpus would be definitely helpful here, but on the other hand it's not easy to generate seed corpora for LPM based fuzz targets. If you can do that (e.g. by writing a helper program that could read actual ASN.1 streams, convert them to proto messages and then dump on disk), that'll be perfect.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+
+namespace asn1_proto {
+
+/* Returns number of bytes needed to encode num
+ * Where num is a variable-length integer with
+ * no leading zeros
+ */
```
I believe most of libfuzzer follows the dominant Google style of `//` and aligned to 78/80 columns.

This is discussed some at https://google.github.io/styleguide/cppguide.html#Comment_Style

We also want to make sure it's a proper sentance, by adding the necessary articles. 
```suggestion
// Returns the number of bytes needed to encode |num| into a variable-length
// two's complement integer with no leading zeros.
```

This is trying to reword the comment so that it's clear what it's encoding *into*.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+ * no leading zeros
+ */
+uint8_t ASN1ProtoConverter::GetNumBytes(size_t num) {
+  uint8_t shift = sizeof(num);
+  while (shift != 0) {
+    if (((num >> ((shift - 1) * 8)) & 0xFF) != 0) {
+      return shift;
+    }
+    shift--;
+  }
```
```suggestion
  for (uint8_t shift = sizeof(num); shift != 0; --shift) {
    if (((num >> ((shift - 1) * 8)) & 0xFF) != 0) {
      return shift;
    }
  }
```

Could be a simpler way to do this.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  return 0;
+}
+
+/* Extracts octets from num and inserts
+ * them in encoder at the position specified
+ */
```
I'm not sure I fully understand this comment. Does this reword mean the same?

```suggestion
// Converts |num| to a variable-length, big-endian representation and inserts
// the result into into |encoder_| at |pos|.
```

That is, function level descriptions are often better in the class file, as with https://google.github.io/styleguide/cppguide.html#Function_Comments

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+size_t ASN1ProtoConverter::EncodeLongForm(const size_t assigned_len,
+                                          const size_t len_pos) {
+  uint8_t len_bytes = GetNumBytes(assigned_len);
+  // long form has hight bit set to 1 and lower bits
+  // set to number of bytes to encode length
```
We want comments to read as complete sentences, per https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

// See X.690 (2015), 8.1.3.5
// Long-form length is encoded as a byte with the high-bit set to indicate the
// long-form, while the remaining bits indicate how many bytes are used to
// encode the length, followed by the actual encoded length.

sleevi commented on projects/asn1-proto/asn1_proto_converter.h:
```diff
@@ -0,0 +1,36 @@
+private:
+  std::stringstream der_;
+  std::vector<uint8_t> encoder_;
+  size_t EncodePDU(const PDU &pdu);
+  size_t EncodeIdentifier(const Identifier &id);
+  uint64_t EncodeHighTagForm(const uint8_t cl, const uint8_t enc, const uint32_t tag);
+  size_t EncodeLength(const Length &len, size_t actual_len, size_t len_pos);
+  size_t EncodeOverrideLength(const std::string len, const size_t len_pos);
+  size_t EncodeIndefiniteLength(const size_t len_pos);
+  size_t EncodeCorrectLength(const size_t actual_len, const size_t len_pos);
+  size_t EncodeLongForm(size_t assigned_len, size_t len_pos);
+  size_t EncodeValue(const Value &val);
+  void AppendBytes(size_t num, size_t pos);
+  uint8_t GetNumBytes(size_t num);
+  void ParseToBits();
```
I think as you move the comments to here, this will be easier to read.

One of the big things slowing me down with this review is constantly trying to make sense of what your "return" indicates. Is it the number of bytes written? Is it the number of bytes that will be written? Something else?

For example, `EncodeLongForm` writes one byte, then returns the number of bytes needed to encode `assigned_len`. But when it's called from `EncodeCorrectLength`, that added long-form byte isn't accounted for, because you return `GetNumBytes(actual_len) + EncodeLongForm(actual_len)` - which seems like an off-by-one?

I still struggle a little with `GetNumBytes`, and I'm wondering if it's better described as `GetVariableIntLen` as `The number of bytes to encode |num| as a variable-length big-endian integer`. For that matter, should `num` be defined as a `uint64_t` instead?

dannyallover replied:
I accepted this change.

dannyallover commented:
> > Have you pushed those changes?
> > I’m concerned about a specific list of tags, as mentioned yesterday, so I’m curious to see what results.
> 
> I did (https://github.com/dannyallover/oss-fuzz/tree/dhalawi/asn1/proto_to_der_2/projects/asn1-proto). Hmm, I wonder why they are not reflecting on this PR.

The changes are there now.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  size_t id_len = EncodeIdentifier(pdu.id());
+  size_t len_pos = encoder_.size();
+  size_t val_len = EncodeValue(pdu.val());
+  size_t len_len = EncodeLength(pdu.len(), val_len, len_pos);
```
I think this has an off-by-one error, per above

dannyallover replied:
Hmm, I do not see the off-by-one error. Could you point me to where you think it is :)?

sleevi replied:
Line 33/34. You don’t account for the leading long-form byte in `EncodeLongForm`, the same way `EncodeIndefiniteLength` didn’t account for the trailing EOC

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  uint8_t enc = static_cast<uint8_t>(id.enc()) << 5;
+  uint32_t tag =
+      id.tag().has_random_tag() ? id.tag().random_tag() : id.tag().known_tag();
+  size_t id_parsed =
```
This should be a uint64_t, right?

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  return len;
+}
+
+size_t ASN1ProtoConverter::EncodeHighTagForm(const uint8_t cls,
```
You changed the header, but forgot to update this to be a uint64_t

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,79 @@
+}
+
+message Identifier {
+  required Class cls = 2;
```
Does this work?

```suggestion
  required Class id_class = 2;
```

(Sorry, every time I see cls I find I'm still struggling reading)

dannyallover replied:
No need to apologize. Same here. That's better than the alternatives I thought of. Change made.

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,79 @@
+
+message Identifier {
+  required Class cls = 2;
+  required Encoding enc = 1;
```
```suggestion
  required Encoding encoding = 1;
```

I don't think encoding is a reserved word?

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+ */
+size_t ASN1ProtoConverter::EncodeIndefiniteLength(const size_t len_pos) {
+  AppendBytes(0x80, len_pos);
+  // value is placed before length
+  // so the pdu's value is already in encoder
+  // we push 0x00 0x00 (End-of-Content) for indefinite form
+  // which is considered a zero-length object so we
+  // need not add anything to the assigned_len
```
Could you reword this comment (and the other comments in the file) to be complete sentences with proper punctuation and grammar, per https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar ?

And also `//` and 78/80 char aligned :)

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  // need not add anything to the assigned_len
+  AppendBytes(0x00, encoder_.size());
+  AppendBytes(0x00, encoder_.size());
+  return 0x01; // it takes one byte to encode 0x80
```
This is part of my confusion about return values. You add 3 bytes, but only return 1 byte (note; you can "`return 1`" to make it a little easier to read)

dannyallover replied:
I was supposed to return 3 bytes. Good catch. The rule is that we always return the number of bytes it takes to encode the thing we insert.

dannyallover replied:
This should return 3. I thought EOC doesn't consume any length: "with the end indicated by the two bytes 00 00 (which can be considered as a zero-length object with tag 0)" (https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/). But I didn't quite understand what that meant until now.

sleevi replied:
right, EOC doesn't affect the length of the current tag (because it's indefinite length), but does affect the length of the parent tag.

For example, if we had
```
SEQUENCE {
  SEQUENCE-with-indefinite-length {
    INTEGER 1
  }
}
```

Then that would be encoded as

```
300730800201010000
```
which is

```
30  -- Sequence
07  -- Length (7 bytes)
  30  -- Sequence
  80  -- Indefinite Length
    02  -- Integer
    01  -- Length (1 byte)
      01  -- Value (1)
  00 -- EOC
  00 -- EOC Length (0)
```

dannyallover replied:
Perfect. That's exactly what I needed to know to clarify.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+
+size_t ASN1ProtoConverter::EncodeValue(const Value &val) {
+  size_t len = 0;
+  if (val.val_array().size() > 0) {
```
Delete this.  The for loop handles this fine on the following line; if there are no elements, it simply doesn't execute the body.

The for loop on 97 basically becomes
`for (const auto& it = val.val_array().begin(); it != val.val_array().end(); ++it)`

and that works out because it == val.val_array.end() when val.val_array().empty(), and so the body isn't run.

dannyallover replied:
If we have ```const auto& it``` then we can't increment it since it's constant. And if we remove the const and have ```auto& it```, then the compiler complains since it's expecting a constant type. I agree that the if statement isn't needed; I believe the current for each loop will not be entered if the size of the vector is zero.

sleevi replied:
Sorry, I re-read this comment and realize it's a bit unclear.

Keep the existing for loop as you have it; you just don't need this conditional. I was trying to explain what your existing for loop expands into (by the compiler), to explain why your existing for loop doesn't need this guard :)

dannyallover replied:
If we have const auto& it, then I get the following error:

> cannot increment value of type 'const google::protobuf::internal::RepeatedPtrIterator<const asn1_proto::ValueElement>'

If I take out the const, I get another error because it's expecting const.

One solution is that the current for each loop will not be entered if the size of the vector is zero, so we can remove the if statement and keep the current loop.

sleevi replied:
Yeah, sorry, I think we collided here. Keep your existing loop, remove the if. The above example is what the compiler expands to.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,166 @@
+  uint8_t numBytes = GetNumBytes(tag);
+  // high tag form requires to set
+  // the lower 5 bits to 1
+  size_t id_parsed = (cls | enc | 0x1F);
```
uint64_t

dannyallover commented:
@sleevi everything has been resolved except for the KnownTag in the proto.

Once I figure out an answer for if we need a KnownTag or not, and once I improve the comments to site the spec directly, I will let you know that I'm ready for another look.

Thanks so much for all the really great feedback :)

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,79 @@
+  Constructed = 1;
+}
+
+message Length {
+  oneof types {
+    bool indefinite_form = 1;
+    bytes length_override = 2;
+  }
+}
```
Let's add some comments here:

```suggestion
message Length {
  // If |types| is absent, the encoded length will be correctly encoded
  // according to {... spec reference here}
  oneof types {
    // If set, causes the length to be encoded using the
    // indefinite-form encoding.
    bool indefinite_form = 1;
    
    // If set, the literal value from |length_override| will be used as the
    // encoded length. This also accounts for incorrectly-encoded
    // lengths (e.g. non-minimal encodings)
    bytes length_override = 2;
  }
}
```

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,79 @@
+
+message Tag {
+  optional uint32 random_tag = 1;
+  required KnownTag known_tag = 2;
```
So, modulo dropping `known_tag`, there's one other element of "invalid DER" we're not covering here, which *might* be useful to cover, rather than having to have it fall back to `val_bits`

```
uint8_t tag_length_override = 3;
```

When `tag_length_override` is set, you treat `Tag` as if it was using the long-form encoding. For example, if `random_tag == 3`, the entire `Identifier` would be encoded within a single byte. However, if `tag_length_override == 3`, the encoded `Identifier` would be `(class | encoding | 0x1F) 0x80 0x80 0x03`.

This is totally *not* valid BER or DER, but some decoders will accept this because it *does* decode as 3, base-128 encoded using the high-tag form.

If you get rid of `KnownTag` and we collapse `Tag` back to just being a `uint32`, then this would go as an option on `Identifier` (e.g. at line 15)

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,151 @@
+  return id_parsed;
+}
+
+uint64_t ASN1ProtoConverter::EncodeIdentifier(const Identifier &id) {
```
er, if EncodeIdentifier returns how many bytes to encode, yeah, it should have been a size_t. Sorry about that.

This is where some of my confusion was coming from, in terms of return values.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,151 @@
+  return GetNumBytes(id_parsed);
+}
+
+size_t ASN1ProtoConverter::EncodePDU(const PDU &pdu) {
```
Note: Because `EncodePDU` calls into `EncodeValue` which calls into `EncodePDU`, you're going to want to keep track of your stack depth to limit your recursion, as covered at https://chromium.googlesource.com/chromium/src/testing/libfuzzer/+/HEAD/libprotobuf-mutator.md#tips-for-grammar-based-fuzzers

dannyallover replied:
Yes, thanks for the reminder! Although, I potentially want to look into better approaches like tail recursion or iterative DFS. But for now, this will do!

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,151 @@
+  size_t id_len = EncodeIdentifier(pdu.id());
+  size_t len_pos = encoder_.size();
+  size_t val_len = EncodeValue(pdu.val());
+  size_t len_len = EncodeLength(pdu.len(), val_len, len_pos);
```
In reviewing this again (and grabbing some coffee), I think it's worth considering what we talked about previously (probably Friday?, possibly further back than that)

That is, because you're basically doing (in highly trash pseudo-code)
```
vector.push_back(tag);
len_offset = vector.size();
vector.push_back(value);
length = vector.size() - len_offset;  // aka the length of what was written
vector.insert(vector.begin() + len_offset, length)
```
You're forcing a `memmove()` every time you encode the length. This almost certainly defeats all of the value you're trying to obtain in avoiding `memcpy()` by not just returning stack-allocated vectors (and, in fact, can be considerably worse due to data access patterns)

What if you did something like:
```
EncodeIdentifier(pdu.id());
EncodeLength(ComputeLength(pdu));
EncodeValue(pdu.val());
```

Note, in this model, you don't need to track return values or offsets at all, because you're always adding to the end.

However, this does require you split out functions for "compute how long something should be" and "encode it and add to the vector". However, this allows you to always-be-appending, avoiding `memmove` entirely. It does mean you take two recursive descents over `pdu`, one to compute the length in `ComputeLength` and the other in `EncodeValue`, but that's far far more efficient than the `memmove` worries.

dannyallover replied:
We talked about this extensively today in our 1:1. The plan is to test what we currently have and then see if a refactor makes sense by looking at things like coverage.

dannyallover replied:
If we don't want to refactor, one option is to do as @davidben suggested and allocate space for the length to avoid frequent `memmove`s. Thanks for the suggestion! Yeah, the way I'm currently doing it is to add in the length after we compute the length of the value.

sleevi commented:
Yeah, the intent is we should be able to use
https://github.com/google/der-ascii/tree/master/cmd/der2ascii except as
der2proto, which can convert all existing corpora. This Protobuf is more or
less interchangeable with that AST.

dannyallover commented:
@sleevi (PTAL) I fixed all the bugs and this now creates valid DER. Additionally, I got to all your feedback, and added comments to cite the spec accordingly.

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  optional uint32 high_tag = 1;
+  // |low_tag| is the default since we know that most of the 
+  // existing ASN1 tags are going to be encoded with one byte.
+  required UInt8 low_tag = 2;
```
Sorry, I realize this probably lost when our chat window lost, but you don't need to go the full range here.

I thought we settled on 

enum LowTag {
  = 0;
  ...
  = 31;
};

That is, we know there's a good number of UNIVERSAL tags in the simple representation (0 .. 2^5-1), and while there are a few UNIVERSAL tags > 31, their appearance "in the real world" is rare and just about equally distributed.

Ultimately, this just helps bias the fuzzer towards the 0...2^5-1 range

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  VAL216 = 216;
+  VAL217 = 217;
+  VAL218 = 218;
+  VAL219 = 219;
+  VAL220 = 220;
+  VAL221 = 221;
+  VAL222 = 222;
+  VAL223 = 223;
+  VAL224 = 224;
+  VAL225 = 225;
+  VAL226 = 226;
+  VAL227 = 227;
+  VAL228 = 228;
+  VAL229 = 229;
+  VAL230 = 230;
+  VAL231 = 231;
+  VAL232 = 232;
+  VAL233 = 233;
+  VAL234 = 234;
+  VAL235 = 235;
+  VAL236 = 236;
+  VAL237 = 237;
+  VAL238 = 238;
+  VAL239 = 239;
+  VAL240 = 240;
+  VAL241 = 241;
+  VAL242 = 242;
+  VAL243 = 243;
+  VAL244 = 244;
+  VAL245 = 245;
+  VAL246 = 246;
+  VAL247 = 247;
+  VAL248 = 248;
+  VAL249 = 249;
+  VAL250 = 250;
+  VAL251 = 251;
+  VAL252 = 252;
+  VAL253 = 253;
+  VAL254 = 254;
+  VAL255 = 255;
```
```suggestion
```

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  required UInt8 low_tag = 2;
+}
+
+enum UInt8 {
```
```suggestion
enum LowTag {
```

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  VAL255 = 255;
+}
+
+// We set class according to X.690, 2015, 8.1.2.2.
```
nit:

```suggestion
// See X.690(2015), 8.1.2.2.
```

I realize my previous comments about full sentences probably lead to this. I think there's definitely a question about whether or not to use pronouns in comments, and this varies among folks, just like tabs vs spaces or emacs vs vim.

I'm a subscriber to the "[Pronouns in comments considered harmful](https://groups.google.com/a/chromium.org/d/msg/chromium-dev/NH-S6KCkr2M/-wTxRAHioN0J)" camp. I've included suggestions throughout, which you can use GitHub's accept suggestion if you'd like, to make it easier for you :)

dannyallover replied:
This is really good to know! Thanks for pointing me to this.

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  Private = 3;
+}
+
+// We set encoding according to X.690, 2015, 8.1.2.5.
```
```suggestion
// See X.690 (2015), 8.1.2.5.
```

sleevi replied:
Note: The style for citing the ITU specs is either `X.690 (08/15)` or `X.690 (2015)`

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+
+message Length {
+  // If |types| is absent, the encoded length will be correctly encoded
+  // according to the DER definite-form (X.690, 2015, 8.1.3-8.1.5 & 10.1).
```
```suggestion
  // according to the DER definite-form (X.690 (2015), 8.1.3-8.1.5 & 10.1).
```

sleevi commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+  // according to the DER definite-form (X.690, 2015, 8.1.3-8.1.5 & 10.1).
+  oneof types {
+    // If |indefinite_form| is set, it causes the length to be encoded using the
+    // indefinite-form encoding (X.690, 2015, 8.1.3.6).
```
```suggestion
    // indefinite-form encoding (X.690 (2015), 8.1.3.6).
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+
+namespace asn1_proto {
+
+// Returns the number of bytes needed to encode |num| into a variable-length
```
Sorry, I realize I haven't been more direct, and this may be causing some confusion:

Documentation like this belongs in the `.h` file, generally. That is, transpose the documentation for each of these to the `.cc` to the `.h`

The effect is that the `.h` can read like the "table of contents / cheat sheet", while the `.cc` file can be the meat of the story :)

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  encoder_.insert(encoder_.begin() + pos, len_vec.begin(), len_vec.end());
+}
+
+// If Override Length is set, then this function will set the length to the
+// arbitrary bytes assigned by the proto.
```
```suggestion
// Writes |raw_length| to |encoder_| at |len_pos| and returns the number of bytes written.
```

The goal here is to describe with the indicative mood, not the imperative mood (see https://google.github.io/styleguide/cppguide.html#Function_Comments )

The other part is trying to document this function as a stand-alone, based purely on the arguments coming in and the arguments going out. The problem with mentioning "override length" is two-fold:
* there's nothing called override length in the inputs/outputs here
* this is describing how the function /is/ used, not how what it does

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+
+// If Override Length is set, then this function will set the length to the
+// arbitrary bytes assigned by the proto.
+size_t ASN1ProtoConverter::EncodeOverrideLength(const std::string len,
```
```suggestion
size_t ASN1ProtoConverter::EncodeOverrideLength(const std::string raw_length,
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  return len.size();
+}
+
+// If Override Length is not set but Indefinite Form is, then this function
+// will set the length and EOC per the Indefinite Form rules (X.690,
+// 2015, 8.1.3.6).
```
```suggestion
// Encodes the indefinite-length indicator (X.690 (2015), 8.1.3.6) at |len_pos|,
// and appends an End-of-Contents (EOC) marker at the end of |encoder_|.
// To correctly call this, the tag must already be encoded immediately prior to
// |len_pos|, and the remainder of |encoder_| represents the encoded value.
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+// 2015, 8.1.3.6).
+size_t ASN1ProtoConverter::EncodeIndefiniteLength(const size_t len_pos) {
+  AppendBytes(0x80, len_pos);
+  // The pdu's value is inserted before the length, so the pdu's value is
+  // already in |encoder_| at the end. Therefore, we push EOC to the end of
+  // |encoder_|.
```
```suggestion
  // The PDU's value is from |len_pos| to the end of |encoder_|, so just add an EOC
  // marker to the end.
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  return 3;
+}
+
+// If Override Length and Inefinite Form are not set, then this function will
+// assign the actual length of the pdu according to DER definite-form (X.690,
+// 2015, 8.1.3-8.1.5 & 10.1).
```
```suggestion
// Encodes |actual_len| using the definite-form length into |encoder_| at
// |len_pos|. See X.690 (2015), 8.1.3-8.1.5 & 10.1.
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+                                               const size_t len_pos) {
+  AppendBytes(actual_len, len_pos);
+  size_t len_num_bytes = GetNumBytes(actual_len);
+  // The long-form is used when the length is larger than 127 (X.690,
+  // 2015, 8.1.3.3).
```
```suggestion
  // X.690(2015), 8.1.3.3: The long-form is used when the length is
  // larger than 127.
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  // The long-form is used when the length is larger than 127 (X.690,
+  // 2015, 8.1.3.3).
+  if (actual_len > 127) {
+    // Long-form length is encoded as a byte with the high-bit set to indicate
+    // the long-form, while the remaining bits indicate how many bytes are used
+    // to encode the length (X.690, 2015, 8.1.3.5).
```
```suggestion
    // See X.690 (2015) 8.1.3.5
    // Long-form length is encoded as a byte with the high-bit set to indicate
    // the long-form, while the remaining bits indicate how many bytes are used
    // to encode the length.
```

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  return len_num_bytes;
+}
+
+size_t ASN1ProtoConverter::EncodeLength(const Length &len,
```
You're still missing documentation (in the `.h` file) for all of these remaining values. Hopefully the above suggestions can provide a model for how to comment. For example

```
// Encodes the length in |actual_len| to DER, returning the length
// in bytes of the encoded length.
// |len_pos| contains the offset in |encoder_| where the length should
// be encoded. |len| can be used to affect the encoding, in order
// to produce invalid lengths.

sleevi commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  }
+}
+
+size_t ASN1ProtoConverter::EncodeValue(const Value &val) {
```
Are you using the `Google` style or the `Chromium` style for the code format? Could you make sure to run with the `Chromium` style? I believe VSCode supports this out-of-the-box.

dannyallover replied:
I was using LLVM. Fixed!

sleevi commented:
Er, I meant to add: This is really coming together, and I think we're in the home stretch, so don't get too discouraged by the number of comments :) 

I've asked @davidben to give a look as well and share his thoughts.

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+// will set the length and EOC per the Indefinite Form rules (X.690,
+// 2015, 8.1.3.6).
+size_t ASN1ProtoConverter::EncodeIndefiniteLength(const size_t len_pos) {
+  AppendBytes(0x80, len_pos);
```
Optional: I think this works, but it's a little odd to be using a big-endian variable-length integer encoder when just encoding a single byte somewhere. I'd probably have inserted individual bytes. (That avoids wondering about things like whether the 0 bytes get encoded as a single 0 byte or the empty string.)

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+// Converts |num| to a variable-length, big-endian representation and inserts
+// the result into into |encoder_| at |pos|.
+void ASN1ProtoConverter::AppendBytes(const size_t num, const size_t pos) {
+  std::vector<uint8_t> len_vec;
```
Nitpick: This function is used for more than just encoding lengths and just generically encodes an integer. It's probably best to have a more generic name than `len_vec`.

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  } else if (len.has_indefinite_form() && len.indefinite_form()) {
+    return EncodeIndefiniteLength(len_pos);
+  } else {
+    return EncodeCorrectLength(actual_len, len_pos);
```
Naming nit: `EncodeCorrectLength` => `EncodeDefiniteLength`? Indefinite lengths are "correct" in BER.

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  // The high tag form base 128 encodes the tag (X.690, 2015, 8.1.2).
+  // We, therefore, calculate how many bytes we need to base 128 encode
+  // the high tag.
+  uint8_t numBytes = GetNumBytes(tag << GetNumBytes(tag));
```
Style: numBytes should be num_bytes.

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  return len;
+}
+
+uint64_t ASN1ProtoConverter::EncodeHighTagForm(const uint8_t id_class,
```
Encoding the integer as a `uint64_t` and then passing it into `InsertBytes` is confusing. I *think* it works, but this is relying on the first byte of the tag being non-zero. Encoding it as bytes directly would be clearer.

dannyallover replied:
I believe the first byte of the tag is always non-zero if it's in high-number-tag form? But I made the change in case I'm wrong about that or if that isn't implicit :)

dannyallover commented:
> Er, I meant to add: This is really coming together, and I think we're in the home stretch, so don't get too discouraged by the number of comments :)
> 
> I've asked @davidben to give a look as well and share his thoughts.

Just got to all your comments -- thank you for the suggestions! Really looking forward to getting @davidben's feedback :)

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+  // The high tag form base 128 encodes the tag (X.690, 2015, 8.1.2).
+  // We, therefore, calculate how many bytes we need to base 128 encode
+  // the high tag.
+  uint8_t numBytes = GetNumBytes(tag << GetNumBytes(tag));
```
`tag << GetNumBytes(tag)` will overflow if `tag` is close to the maximum value. It also doesn't actually count the number of base 128 words needed, though it does happen to work for all values of `tag` that fit.

But, for instance, n = 2^64-1 is 64 bits long. GetNumBytes(n) is 8. n << 8 is 72 bits long, so this would (ignoring overflow) return 9. However, n needs 10 digits in base 128 representation. (ceil(64/7) = 10). I would suggest just writing a loop like `GetNumBytes` but counting by 7 instead of 8.

This is what I did in der-ascii. (Ignore the length override. That's to support a length override feature.)
https://github.com/google/der-ascii/blob/master/cmd/ascii2der/encoder.go#L33

dannyallover replied:
Ah, I didn't notice that! Another alternative is to make base a function parameter in GetNumBytes, since the two functions will be identical except for the shift.

davidben replied:
Ah yeah, that would work. You could also count bits and return `ceil(num_bits / base_bits)` or `(num_bits + base_bits - 1) / base_bits`.

davidben commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+}
+
+size_t ASN1ProtoConverter::EncodeIdentifier(const Identifier &id) {
+  // The class comprises the the 7th and 8th bit of the identifier (X.690,
```
"the the" => "the"

dannyallover commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+}
+
+message Length {
+  // If |types| is absent, the encoded length will be correctly encoded
```
**Alternative One:**
```
message Length {
   optional bytes length_override = 1;
   required bool indefinite_form = 2;
```

```
  if (len.has_length_override()) {
    return EncodeOverrideLength(len.length_override(), len_pos);
  } else if (len.indefinite_form()) {
    return EncodeIndefiniteLength(len_pos);
  } else { // use the definite form
    return EncodeDefiniteLength(actual_len, len_pos);
```

**Alternative Two:**
We keep the proto as is but change the following:
```
  if (len.has_length_override()) {
    return EncodeOverrideLength(len.length_override(), len_pos);
  } else if (len.has_indefinite_form()) {
    return EncodeIndefiniteLength(len_pos);
  } else { // use the definite form
    return EncodeDefiniteLength(actual_len, len_pos);
```
Here, we need not check if indefinite_form is true; essentially if it is present we interpret it as use indefinite form.

You bring up a very good point here! Yes, unfortunately, protobuf does not have a unit type :/, and we've been using work arounds like this (https://chromium.googlesource.com/chromium/src/testing/libfuzzer/+/HEAD/libprotobuf-mutator.md#tips-for-grammar-based-fuzzers). Above are two possible work arounds. I was thinking though that this might not be an issue because the double check on indefinite form makes it less likely for us to use the indefinite form, and in most cases we will want to use the actual length in order to get deep into the structure?

dannyallover replied:
I was thinking though that this might not be an issue because the double check on indefinite form makes it less likely for us to use the indefinite form, and in most cases we will want to use the actual length in order to get deep into the structure? Let me know if I'm missing something here :)

davidben replied:
Wouldn't the second cause the indefinite form to be chosen more frequently? But that's a good point; we care both about redundancy and frequency (e.g. your LowTag enum is redundant but seems still useful).

I've never actually worked libprotobuf-mutator, so I don't know how best to optimize these structures. Perhaps this is something you figure out by observing the fuzzer's coverage? Anyway, I don't know how this works, so I'm happy with any of the three (existing definition, alternative 1, and alternative 2). :-)

dannyallover commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+
+message ValueElement {
+  optional PDU pdu = 1;
+  required bytes val_bits = 2;
```
The only issue with the oneof is that (as you pointe dout) `oneof` is really `atmostoneof` not `exactlyoneof.` So, it's not clear what we would insert if it's empty.

dannyallover replied:
I believe we would need a required, optional, or repeated in front of the PDU?

sleevi replied:
Ah, right, yes, `required`. I forgot proto2 mandates one of those values :)

dannyallover commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,319 @@
+
+    // If |length_override| is set, the literal value from |length_override|
+    // will be used as the encoded length. This also accounts for 
+    // incorrectly-encoded lengths (e.g. non-minimal encodings).
```
Yes, incorrect here, meaning incorrect DER. I made that explicit.

dannyallover commented:
> I ended up looking at the version of the PR before you addressed Ryan's comments, so apologies if some of mine are redundant now.
> 
> Very excited to see this coming together, and welcome to the ASN.1 implementers' club. It is a... gnarly format. :-)

Your comments didn't overlap with @sleevi 's so no worries there! I just made all the appropriate changes, and want to thank you for the very helpful feedback :)

@sleevi PTAL and let me know if you see any room for additional improvements :)

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,153 @@
+  }
+  id_parsed |= (tag_num & 0x7F);
+  AppendBytes(id_parsed, encoder.size());
+  return GetNumBytes(id_parsed, 256);
```
Is this right/necessary? Can't you just return `num_bytes + 1` ? 

That is, you know you'll always have at least one byte (because of line 89), and then you add enough bytes to represent the tag, which is `num_bytes`. This avoids needing to re-read/re-compute the length

dannyallover replied:
To answer your question, it is right. But what you suggested is simpler + more efficient. Change accepted!

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,153 @@
+
+uint8_t ASN1ProtoToDER::GetNumBytes(const size_t num, const size_t base) {
+  uint8_t base_bits = log2(base);
+  for (uint8_t num_bits = sizeof(num) * 8; num_bits != 0; --num_bits) {
+    if ((num >> num_bits) & 0x01) {
+      return ceil(num_bits / base_bits);
+    }
+  }
```
```suggestion
  for (uint8_t num_bits = sizeof(num) * 8; num_bits > base_bits; num_bits -= base_bits) {
    if ((num >> num_bits)) {
      return ceil(num_bits / base_bits);
    }
  }
```

You don't need the mask here, because any bit being 1 counts.

The further optimization is to check `base_bits` worth of bits at a time (i.e. either 7 or 8 bits). This is in effect 'loop unrolling'.

For example, for a base-128 number, it's going to do:
base_bits = 7
num_bits = 64
num_bits = 57
num_bits = 50
num_bits = 43
num_bits = 36
num_bits = 29
num_bits = 22
num_bits = 15
num_bits = 8
num_bits = 1  // terminate the loop

It doesn't matter if the last bit is set or not because we know both 0 and 1 require 1 byte to represent.

Although you probably want to spot-check my math here :)

dannyallover replied:
Ah, very nice, change accepted!

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,153 @@
+}
+
+size_t ASN1ProtoToDER::EncodePDU(const PDU& pdu) {
+  depth_++;
```
```suggestion
  ++depth_;
```

Dor1s commented:
> I'd love to hear from @Dor1s on where we might be able to place this (e.g. should we do a separate repository, ala LPM) to allow individual projects to check it out as part of their fuzz integration.

I think we should create a `proto` directory in https://github.com/google/fuzzing, then create an `asn1` subdir in there and put the following files inside: `asn1.proto`, `asn1_proto_to_der.cc `, `asn1_proto_to_der.h `. Additionally may also include a README with any other content.

Then in OSS-Fuzz (e.g. for boringssl project):

1) add git clone google/fuzzing repo in https://github.com/google/oss-fuzz/blob/master/projects/boringssl/Dockerfile
2) add fuzz target source in https://github.com/google/oss-fuzz/tree/master/projects/boringssl
3) add build commands to https://github.com/google/oss-fuzz/blob/master/projects/boringssl/build.sh

Ideally, step 2) should go into the boringssl repo, but I feel like we may want to work more with it inside OSS-Fuzz repo first and maybe later move it into boringssl.

dannyallover commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,95 @@
+
+message TagNumber {
+  optional uint32 high_tag_num = 1;
+  // |low_tag| is the default since we know that most of the 
+  // existing ASN1 tags are going to be encoded with one byte.
```
I just want to be clear here. When we say "most tags" are we talking about the class+number or are we talking about just the number. 

For me, `LowTagNumber` refers to the tag number, and therefore, I believe that the comment should be changed to something like the following:
// |low_tag_number| is the default since we know that most of the 
// existing ASN1 tags numbers are going to be encoded with one byte.

sleevi replied:
We're specifically talking about the tag number, and this holds true across all class types, that their tag number is designed to fit within a 1-byte identifier.

Note: I'm still trying to make sure we avoid "we" comments as much as possible, because of ambiguity.

```c++
// Most tags will fit within the range of a single byte Identifier
// (meaning the tag is within the range 0..30), so make that the
// default.
```

Thanks for pointing out my awkward wording suggestion, good catch 👍

dannyallover replied:
No problem. Just one final touch on your suggested comment (to make it very explicit). Let me know if you agree.

```// Most tags will fit within the range of a single byte Identifier
// (meaning the **tag number** is within the range 0..30), so make that the
// default.```

dannyallover commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,153 @@
+
+void ASN1ProtoToDER::AppendBytes(const size_t num, const size_t pos) {
+  std::vector<uint8_t> len_vec;
+  for (uint8_t shift = GetNumBytes(num, 256); shift != 0; shift--) {
```
@sleevi please let me know if you would like me to read through the style guide to avoid future style errors!

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,156 @@
+}
+
+size_t ASN1ProtoToDER::EncodeIndefiniteLength(const size_t len_pos) {
+  AppendVariableInt(0x80, len_pos);
+  // The PDU's value is from |len_pos| to the end of |encoder_|, so just add an
+  // EOC marker to the end.
+  AppendVariableInt(0x00, encoder_.size());
+  AppendVariableInt(0x00, encoder_.size());
```
With the new method name, it becomes easier to spot things that don't look right. For example, here, we're not really wanting to append some variable integer: we want one and exactly one byte.

```suggestion
  encoder_.insert(encoder_.begin() + len_pos, 0x80);
  // The PDU's value is from |len_pos| to the end of |encoder_|, so just add an
  // EOC marker to the end.
  encoder_.push_back(0x00);
  encoder_.push_back(0x00);
```

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,156 @@
+    // Long-form length is encoded as a byte with the high-bit set to indicate
+    // the long-form, while the remaining bits indicate how many bytes are used
+    // to encode the length.
+    AppendVariableInt((0x80 | len_num_bytes), len_pos);
```
Similarly, the rename helps highlight that we're not using a var int here:

```suggestion
    encoder_.insert(encoder_.begin() + len_pos, (0x80 | len_num_bytes));
```

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,156 @@
+  if (tag_num >= 31) {
+    return EncodeHighTagNumberForm(id_class, encoding, tag_num);
+  }
+  AppendVariableInt((id_class | encoding | tag_num), encoder_.size());
```
```suggestion
  encoder_.push_back(static_cast<uint8_t>(id_class | encoding | tag_num));
```

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
How'd you come up with this number, fwiw? I think something like 200 is probably plenty sufficient?

```suggestion
  if (depth_ > kRecursionLimit) {
```

dannyallover replied:
Just by forcing a bunch of recursive calls and seeing how far we can go, then cutting off the number by a constant to be safe. But yeah you're right, we don't need that many! Change made.

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
I'm assuming this debug code will be removed before landing? :)

dannyallover replied:
Yes, just removed :)

sleevi commented on projects/asn1-proto/asn1_proto_to_der.h:
```diff
@@ -0,0 +1,80 @@
+  // Prints bits in |encoder_|. Used for testing and validation.
+  void PrintEncodedBits();
+  
+  size_t depth_;
```
```suggestion
  // Tracks recursion depth to avoid stack exhaustion.
  size_t depth_;
```

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,156 @@
+#include "asn1_proto_to_der.h"
+
+namespace asn1_proto {
+
```
```suggestion
// The maximum level of recursion allowed. Values greater than this will just fail.
static constexpr size_t kRecursionLimit = 67000;

// Return value used to signal that the recursion limit has been exceeded by a
// message or value and to abort further processing.
static constexpr size_t kErrRecursionLimitReached = SIZE_MAX;
```

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```diff
@@ -0,0 +1,156 @@
+  size_t len = 0;
+  for (const auto& val_ele : val.val_array()) {
+    if (val_ele.has_pdu()) {
+      len += EncodePDU(val_ele.pdu());
```
```suggestion
      size_t pdu_len = EncodePDU(val_ele.pdu());
      if (pdu_len == kErrRecursionLimitReached) {
        // If the recursion limit was exceeded, fail processing
        // the entire message, to avoid uninteresting inputs.
        return kErrRecursionLimitReached;
      }
      len += pdu_len;
```

dannyallover replied:
I no longer want the Encode helpers to return anything, so I changed kErrRecursionLimitReached to a bool that I set as true when the limit is reached and check for that.

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```suggestion
  size_t val_len = EncodeValue(pdu.val());
  if (val_len == kErrRecursionLimitReached) {
    return kErrRecursionLimitReached;
  }
```

dannyallover replied:
I no longer want the Encode helpers to return lengths, so I changed kErrRecursionLimitReached to a bool that I set as true when the limit is reached and check for that.

sleevi commented on projects/asn1-proto/asn1_proto_to_der.cc:
```suggestion
  size_t encoded_len = EncodePDU(pdu);
  if (encoded_len == kErrRecursionLimitReached) {
    encoder_.clear();
  }
```

dannyallover replied:
I don't think this can happen since ProtoToDER is called on the parent PDU, and when that occurs we will have made zero function calls. Let me know if I'm missing something!

sleevi replied:
The idea was to propagate the error code upwards through the stack, so that you don’t encode any message whatsoever if the message would exceed the recursion depth.

dannyallover replied:
Ah, I see. That makes sense!

dannyallover commented on projects/asn1-proto/asn1_proto_converter.cc:
```diff
@@ -0,0 +1,169 @@
+
+// Converts |num| to a variable-length, big-endian representation and inserts
+// the result into into |encoder_| at |pos|.
+void ASN1ProtoConverter::AppendBytes(const size_t num, const size_t pos) {
```
I missed this comment; yes I totally agree. I changed it to the following:

```
// Converts |value| to a variable-length, big-endian representation and inserts
// the result into into |encoder_| at |pos|.
void InsertVariableInt(const size_t value, const size_t pos);
```

Let me know if that still doesn't look clear!

dannyallover commented on projects/asn1-proto/asn1_proto_converter.cc:
Yes, @sleevi suggested this as well. I think that will improve simplicity/readability and avoid potential bugs in returning the wrong lengths in some of the helper functions. Change made!

dannyallover commented on projects/asn1-proto/asn1_proto_to_der.h:
```diff
@@ -0,0 +1,68 @@
+ private:
+  size_t depth_;
+  std::vector<uint8_t> encoder_;
+  // Enocdes |pdu| to DER, returning the number of bytes needed encode |pdu|.
```
This is a good idea! It allows us to avoid bugs (since a lot of bugs were coming from the size issue), and will allow us to simplify a lot of the methods. I made the change :)

dannyallover commented on projects/asn1-proto/asn1.proto:
```diff
@@ -0,0 +1,96 @@
+  VAL28 = 28;
+  VAL29 = 29;
+  VAL30 = 30;
+  VAL31 = 31;
```
Good catch!

dannyallover commented on projects/asn1-proto/asn1_proto_converter.h:
```diff
@@ -0,0 +1,36 @@
+
+class ASN1ProtoConverter {
+public:
+  std::vector<uint8_t> ProtoToDER(const PDU &pdu);
```
I agree that I misbehaves if we call `ProtoToDER` twice on the same `ASN1ProtoConverter`. I just fixed that. Could you please provide more detail on this "Perhaps just a freefloating function and the helper state can just be an implementation detail inside the cc file?". I think I agree with you, but I just want to make sure I follow :)

dannyallover commented:
@sleevi @davidben @Dor1s Thank you all so much for the very fruitful feedback. I realized today that a missed some of the comments, so I just did a double check and got to the rest. I'm going to be submitting a PR tomorrow to https://github.com/google/fuzzing were this ASN1 Fuzzer will live. Looking forward to getting more feedback there :)
